http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/AwsCrtAsyncHttpClient.java

```diff
@@ -0,0 +1,370 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.crt.CrtResource;
+import software.amazon.awssdk.crt.CrtRuntimeException;
+import software.amazon.awssdk.crt.http.HttpClientConnectionManager;
+import software.amazon.awssdk.crt.http.HttpClientConnectionManagerOptions;
+import software.amazon.awssdk.crt.http.HttpHeader;
+import software.amazon.awssdk.crt.http.HttpRequest;
+import software.amazon.awssdk.crt.io.ClientBootstrap;
+import software.amazon.awssdk.crt.io.EventLoopGroup;
+import software.amazon.awssdk.crt.io.HostResolver;
+import software.amazon.awssdk.crt.io.SocketOptions;
+import software.amazon.awssdk.crt.io.TlsCipherPreference;
+import software.amazon.awssdk.crt.io.TlsContext;
+import software.amazon.awssdk.crt.io.TlsContextOptions;
+import software.amazon.awssdk.http.SdkHttpClient;
+import software.amazon.awssdk.http.SdkHttpConfigurationOption;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.http.async.AsyncExecuteRequest;
+import software.amazon.awssdk.http.async.SdkAsyncHttpClient;
+import software.amazon.awssdk.http.crt.internal.AwsCrtAsyncHttpStreamAdapter;
+import software.amazon.awssdk.utils.AttributeMap;
+import software.amazon.awssdk.utils.IoUtils;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Validate;
+import software.amazon.awssdk.utils.http.SdkHttpUtils;
+
+/**
+ * An implementation of {@link SdkHttpClient} that uses the AWS Common Runtime (CRT) Http Client to communicate with
+ * Http Web Services. This client is asynchronous and uses non-blocking IO.
+ *
+ * <p>This can be created via {@link #builder()}</p>
+ */
+@SdkPublicApi
+public class AwsCrtAsyncHttpClient implements SdkAsyncHttpClient {
+    private static final Logger log = Logger.loggerFor(AwsCrtAsyncHttpClient.class);
+    private static final String HOST_HEADER = "Host";
+    private static final String CONTENT_LENGTH = "Content-Length";
+    private static final String CONNECTION = "Connection";
+    private static final String KEEP_ALIVE = "keep-alive";
+    private static final String AWS_COMMON_RUNTIME = "AwsCommonRuntime";
+    private static final int DEFAULT_STREAM_WINDOW_SIZE = 16 * 1024 * 1024; // 16 MB
+
+    private final Map<URI, HttpClientConnectionManager> connectionPools = new ConcurrentHashMap<>();
+    private final LinkedList<CrtResource> ownedSubResources = new LinkedList<>();
+    private final AtomicBoolean isClosed = new AtomicBoolean(false);
+    private final ClientBootstrap bootstrap;
+    private final SocketOptions socketOptions;
+    private final TlsContextOptions tlsContextOptions;
+    private final TlsContext tlsContext;
+    private final int windowSize;
+    private final int maxConnectionsPerEndpoint;
+
+
+    public AwsCrtAsyncHttpClient(DefaultBuilder builder, AttributeMap config) throws CrtRuntimeException {
```

Comment:
Are you sure we want to declare a checked exception here?
caveat: I know very little about standards/style with Java exceptions

Reply:
I originally pulled that over as part of inlining the ClientBootstrap constructor that's since been removed.  (It created the EventLoopGroup and HostResolver like I had also previously inlined.)  Because they are now passed as arguments, and nothing is being added, I don't think anything is being added here that warrants in.  (This is not to say that it *had* to be brought over in the first place.). So, I have taken it out.  However, I would be open to adding it back in if there is a good argument for it!